feat(#59): 주문 생성 API 구현

joel-you opened the pull request:
## 📋 PR 요약

주문 생성 API의 검증 로직 강화, 테스트 커버리지 확보, API 문서화 작업을 완료했습니다.

### 🔗 관련 이슈

- #59

### 📝 변경 사항

- 새로운 기능 추가
- 리팩토링
- 문서 업데이트
- 테스트 코드 추가/수정

### 🛠️ 기술적 변경 사항

수정된 파일:
- OrderItemRequest.java: unitPrice 검증을 @Min에서 @DecimalMin(0.01)로 개선

새로 추가된 파일:
- OrderControllerTest.java: Controller 통합 테스트 6개 (성공 1개, 실패 5개)
- OrderServiceTest.java: Service 단위 테스트 1개
- order-orchestrator-openapi.yml: OpenAPI 3.0.3 스펙 문서

주요 로직 변경:
- BigDecimal 타입에 적합한 검증 어노테이션 적용으로 0원 상품 방지

### 🧪 테스트

- 단위 테스트 통과
- 통합 테스트 통과

테스트 커버리지: Line 100% (11/11), Branch 66.67% (2/3)

테스트 케이스:
1. 정상 주문 생성 (성공)
2. customerId null 검증 (실패)
3. orderItems 빈 배열 검증 (실패)
4. unitPrice 0원 검증 (실패)
5. quantity 0 이하 검증 (실패)
6. 잘못된 JSON 형식 검증 (실패)

### 📖 문서 업데이트

- API 문서 업데이트 완료 (OpenAPI 스펙, Redocly 검증 완료)
- 코드 주석 업데이트 완료

### ✅ 체크리스트

- 코드가 컨벤션을 따르고 있습니다
- 자가 리뷰를 완료했습니다
- 테스트를 추가/업데이트했습니다
- CI/CD 파이프라인이 통과합니다

joel-you committed:
feat(#59): 주문 생성 API 구현

- Validation이 적용된 DTO 추가 (Request/Response)
- UseCase 및 Command 패턴 구현
- 주문 번호 자동 생성 기능이 포함된 OrderService 추가
- REST 엔드포인트 POST /api/v1/orders 추가
- MapStruct DTO 매퍼 추가

joel-you committed:
Merge branch 'feature/58-order-db-schema-and-jpa-repository' into feature/59-order-create-api

joel-you committed:
refactor(#59): 주문 상품 단가 검증 로직 개선

- @Min을 @DecimalMin으로 변경하여 소수점 단위 가격 지원
- 최소 가격을 0.01로 설정하여 0원 상품 방지
- BigDecimal 타입에 적합한 검증 어노테이션 적용

joel-you committed:
test(#59): 주문 생성 API 테스트 작성

- OrderControllerTest: POST /api/v1/orders 성공 케이스 검증
- OrderServiceTest: 주문 생성 로직 단위 테스트
- MockMvc 기반 통합 테스트로 실제 HTTP 요청/응답 검증
- 테스트 커버리지 70% 이상 달성

joel-you committed:
docs(#59): Order Orchestrator OpenAPI 스펙 문서 작성

- POST /api/v1/orders 엔드포인트 명세 작성
- 요청/응답 스키마 정의 (CreateOrderRequest, OrderResponse)
- 에러 응답 스키마 정의 (ErrorResponse, ValidationError)
- 비즈니스 규칙 및 검증 규칙 상세 문서화
- 다양한 예제 시나리오 포함 (성공/실패 케이스)

joel-you committed:
test(#59): 주문 생성 API 검증 실패 케이스 테스트 추가

- customerId null 검증 테스트
- orderItems 빈 배열 검증 테스트
- unitPrice 0원 검증 테스트
- 검증 실패 시 UseCase 미호출 확인

joel-you committed:
test(#59): 주문 생성 API 추가 엣지 케이스 테스트

- quantity 0 이하 검증 테스트 추가
- 잘못된 JSON 형식 처리 테스트 추가
- API 엣지 케이스 커버리지 강화

joel-you committed:
refactor(#59): 주문 응답 객체 계층 분리 및 매퍼 테스트 커버리지 개선

- OrderResponse, OrderItemResponse를 application layer로 이동
- OrderDtoMapper를 OrderMapper로 리팩토링 (request-to-command 변환)
- CreateOrderUseCase가 OrderResponse 직접 반환하도록 변경
- OrderController 로직 간소화 (ProductController 스타일)
- OrderResponseMapper 추가 (domain-to-response 변환)
- 테스트 커버리지 개선
  * monolith: ProductMapperTest 추가 (55% 달성)
  * order-orchestrator: OrderMapperTest 추가 (70% 달성)

joel-you committed:
refactor(#65): PR 리뷰 반영 - 인덱스 최적화 및 JPA 개선

- 복합 인덱스 우선 설계로 불필요한 인덱스 제거
- JPQL을 Spring Data JPA 메소드로 변환
- JOIN FETCH를 EntityGraph로 변경

joel-you committed:
refactor(#58): UuidGenerator를 common 모듈로 이동 및 라이브러리 변경

- UUID 라이브러리를 Java UUID Generator(FasterXML)로 변경
- buildSrc에 UUID Generator 의존성 추가하여 중앙 집중식 관리
- UuidGenerator를 common 모듈로 이동하여 모든 서비스에서 재사용 가능하게 개선
- Flyway 마이그레이션 스크립트 에러 수정 (FK, AUTO_INCREMENT, 인덱스 제거 로직 보완)

joel-you committed:
refactor(#65): OrderRepositoryImpl save 메서드 단순화

- update/insert 분기 로직 제거, insert 전용으로 변경
- syncOrderItems 차등 동기화 로직 제거

joel-you committed:
refactor(#65): V1 마이그레이션 SQL 중복 코드 제거

- 반복되는 인덱스 제거 로직을 stored procedure로 리팩토링
- drop_index_if_exists 프로시저 도입으로 코드 재사용
- SonarCloud 중복 코드 경고 해결

joel-you committed:
refactor(#58): 마이그레이션 SQL 코드 품질 개선

- 테이블명 SQL 변수로 상수화
- UPDATE order_items에 WHERE 조건 추가
- SonarQube Critical 이슈 3건 해결

joel-you committed:
Merge remote-tracking branch 'origin/feature/58-order-db-schema-and-jpa-repository' into feature/59-order-create-api

joel-you committed:
refactor(#59): UUID v7을 orderNumber로 직접 사용하도록 개선

joel-you commented:
수정완료

joel-you committed:
Remove .github/README.md from feature/59-order-create-api

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/application/service/OrderService.java:
```diff
@@ -0,0 +1,62 @@
+    public OrderResponse createOrder(CreateOrderCommand command) {
+        String orderNumber = UuidGenerator.generate().toString();
+
+        Map<String, Object> defaultShippingAddress = new HashMap<>();
```
Map<String, Object> defaultShippingAddress는 타입 안정성에 취약할 것 같아요.
`ShippAddress(VO)`와 `AddressType(enum)` 조합을 사용하는 방향은 어떠세요?

joel-you replied:
이게 더 좋을 것 같네요. 수정하겠습니다!

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/application/service/OrderService.java:
```diff
@@ -0,0 +1,62 @@
+        command.orderItems().forEach(itemCommand -> {
+            OrderItem orderItem = OrderItem.create(
+                itemCommand.productId(),
+                "Product-" + itemCommand.productId(),
```
`prefix`가 필요한 상황인지 궁금합니다.
조회 용도에서 필요한 것이라면 도메인에서 선택적으로 정보를 다루는 방향이 좋을 것 같아요.

joel-you replied:
필요 없을 것 같네요. 제거하겠습니다

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/adapter/in/web/dto/OrderItemRequest.java:
```diff
@@ -0,0 +1,24 @@
+import lombok.Builder;
+
+@Builder
+public record OrderItemRequest(
```
여기에서는 문법적으로 에러 유무를 확인하고 있는데요. 실제 `의미`/`정합성에` 대한 검증도 필요할 것 같습니다. 
클라이언트에서 보낸 unitPrice가 불일치 시 정책, 재고에 따른 구매 한도 등 UseCase 에서 정합성에 대한 검증을 고민해 주시면 좋겠습니다.

flab-true commented on order-orchestrator/src/main/java/com/msa/commerce/orchestrator/application/service/OrderService.java:
```diff
@@ -0,0 +1,62 @@
+    private final OrderResponseMapper orderResponseMapper;
+
+    @Override
+    public OrderResponse createOrder(CreateOrderCommand command) {
```
OrderItemRequest 코멘트에서 작성한 외부 시스템 기준의 정합성이 필요할 것 같습니다. 
본 아키텍처에서 `Domain Model` 에서 검증을 할 것인지 `Application Service` 에서 검증할 것인지 고민을 해주시고, 
`UseCase` 내부에서 `Port`를 통해 정책 검증을 고려해 주세요.

flab-true reviewed (commented):
코멘트 남깁니다.

flab-true commented:
헥사고날 아키텍처에서 `도메인 계층(domain model)`과 `UseCase(Application Service)`에 대한 온도감을 맞추면 좋겠습니다. 
- `UseCase(Application Service)`  에서 `도메인 로직을 오케스트레이션`
- 외부 시스템과 협력, 정책 검증 방안 등의

검증 종류 | 위치 | 예시 
-- | -- | -- 
형식 검증 | Controller (입력 DTO) | @NotNull, @Min, @Pattern 
정책 검증 | UseCase (Application) | 가격 일치, 재고 유효 
불변식 검증 | Domain(Entity/VO) | 수량>0, 합계 금액 등

joel-you committed:
Merge branch 'develop' into feature/59-order-create-api

Resolved conflicts

joel-you committed:
refactor(#59): 배송지 주소를 위한 Value Object 추가

- AddressType enum 추가 (DEFAULT, ALTERNATE, GIFT)
- ShippingAddress record 추가 (불변 객체)
- 타입 안정성을 위해 Map<String, Object> 대신 VO 사용
- 정적 팩토리 메서드 제공 (createDefault, create)

joel-you committed:
refactor(#59): Order 엔티티 배송지 타입을 VO로 변경

- Map<String, Object>를 ShippingAddress로 변경
- getShippingAddress() 메서드 제거 (Lombok getter 자동 생성)

joel-you committed:
refactor(#66): Value Object 및 Port 패턴 도입으로 타입 안정성 및 아키텍처 개선

- ProductPort 인터페이스로 외부 서비스 의존성 추상화
- ProductInfo VO로 상품 정보 불변 객체화
- MockProductAdapter 임시 구현체 추가
- 하드코딩된 "Product-" 접두사 제거
- MapStruct 매퍼에 VO ↔ Map 변환 로직 추가

joel-you committed:
refactor(#59): RestClient 기반 Product Service 통합 및 Exception 아키텍처 개선

- EKS Service Name을 활용한 Kubernetes 서비스 디스커버리 지원
- RestClientFactory 및 Properties를 common 모듈로 추출하여 재사용성 향상
- Product 도메인 특화 Exception을 monolith 모듈로 이동
- 사용되지 않는 NoChangesProvidedException 제거
- .gitignore 수정으로 out 패키지 경로 문제 해결
